### PR TITLE
WIP Add size to abstract StaticArray type

### PR DESCRIPTION
**Pull request by andyferris**
This is a continuation of #127 whereby we add a `Tuple{...}` size descriptor to the abstract `StaticArray` type.

I'm previewing this WIP to get feedback on the interface, such as whether we want `StaticArray{Size, T, N}`, or have the dimensions after `T`, or after `N`. Any comments appreciated, otherwise I'll carry on.

### CODE REVIEW

**Comment by andyferris** [comment]
OK, bulk of work is done, but I haven't updated the tests so many are failing.

More alarmingly, I'm running into an inference crash in the unit tests; similar things have been reported [elsewhere](https://github.com/JuliaLang/julia/issues/21244).

**Comment by andyferris** [comment]
This is ready for a look at. There is a problem with inference (world age failure) in a testset for solve, but it makes no sense to me.

I would still like some feedback on design here, please. :)

**Comment by c42f** [comment]
Afraid I don't have the energy to look at this tonight... maybe tomorrow.

File: test/solve.jl [hunk 1]
```diff
@@ -1,11 +1,27 @@
 @testset "Solving linear system" begin
 
-    @testset "Problem size: $n x $n. Matrix type: $m. Element type: $elty" for n in (1,2,3,4),
+    # I see an inference error in the combined testset below. I can't reproduce
+    # at the REPL or in a function for individual n, etc... speculatively, it
+    # might be reusing A, b with different types in the same (non-toplevel)
+    # scope, for which I've come accross inference bugs in the past.
```
Comment (c42f) [inline test/solve.jl]: Somewhat disturbing.

Seems mitigated by the fact that this test case is rather odd with its loop over types though.  May not be a problem in practice?

File: src/SArray.jl [hunk 2]
```diff
@@ -1,72 +1,70 @@
 """
-    SArray{Size, T, L}(x::NTuple{L, T})
-    SArray{Size, T, L}(x1, x2, x3, ...)
+    SArray{S, T, L}(x::NTuple{L, T})
+    SArray{S, T, L}(x1, x2, x3, ...)
```
Comment (andyferris) [inline src/SArray.jl]: This I always find a bit weird. In some sense, we are documenting the type *and* the constructors. The one without `T` and `L` is documented below.

The `L` is necessary to know about whenever you want to specify it as an element type of an array or a member of a struct, so this was my effort to get the user to find out about it's existence.

Though now I notice that I forgot the `N`!

File: src/SMatrix.jl [hunk 3]
```diff
@@ -53,9 +53,12 @@ end
     end
 end
 
-@inline convert{S1,S2,T}(::Type{SMatrix{S1,S2}}, a::StaticArray{T}) = SMatrix{S1,S2,T}(Tuple(a))
+@inline convert{S1,S2,T}(::Type{SMatrix{S1,S2}}, a::StaticArray{<:Any, T}) = SMatrix{S1,S2,T}(Tuple(a))
```
Comment (andyferris) [inline src/SMatrix.jl]: It does allow reshaping... but primarily it provides the functionality to change `T`.

Should we tighten this up at some point?

File: src/SUnitRange.jl [hunk 4]
```diff
@@ -1,8 +1,23 @@
 # `Start` and `End` should be `Int`
-struct SUnitRange{Start,End} <: StaticVector{Int}
+struct SUnitRange{Start, End, L} <: StaticVector{L, Int}
```
Comment (andyferris) [inline src/SUnitRange.jl]: I guess we could - user's can use the `@pure` constructor on line 20.

File: src/SizedArray.jl [hunk 5]
```diff
@@ -93,4 +84,4 @@ Creates a `SizedArray` wrapping `array` with the specified statically-known
 `dims`, so to take advantage of the (faster) methods defined by the static array
 package.
 """
-(::Size{S}){S}(a::Array) = SizedArray{S}(a)
+(::Size{S}){S}(a::Array) = SizedArray{Tuple{S...}}(a)
```
Comment (andyferris) [inline src/SizedArray.jl]: True. Make a PR :)

File: src/abstractarray.jl [hunk 6]
```diff
@@ -39,32 +38,26 @@ if they wish to overload the default behavior.
...
 
 # Default types
-# Generally, use SVector, etc
+# Generally, use SArray
 similar_type{A<:AbstractArray,T,S}(::Type{A},::Type{T},s::Size{S}) = default_similar_type(T,s,length_val(s))
-
-default_similar_type{T,S}(::Type{T}, s::Size{S}, ::Type{Val{0}}) = Scalar{T}
-default_similar_type{T,S}(::Type{T}, s::Size{S}, ::Type{Val{1}}) = SVector{S[1],T}
-default_similar_type{T,S}(::Type{T}, s::Size{S}, ::Type{Val{2}}) = SMatrix{S[1],S[2],T,prod(s)}
 default_similar_type{T,S,D}(::Type{T}, s::Size{S}, ::Type{Val{D}}) = SArray{Tuple{S...},T,D,prod(s)}
```
Comment (andyferris) [inline src/abstractarray.jl]: The length of a `Scalar` is `1`.

Try `SArray{Tuple{},Int,0,1}((3,))`

Comment (andyferris) [inline src/abstractarray.jl]: Also `prod([]) === 1`

File: test/solve.jl (hunk 1 above)
Comment (andyferris) [inline test/solve.jl]: I think it's not a problem in practice, or at least I couldn't find any evidence of a problem.

File: src/SMatrix.jl (hunk 3 above)
Comment (andyferris) [inline src/SMatrix.jl]: (perhaps a separate PR)

File: src/SArray.jl (hunk 2 above)
Comment (c42f) [inline src/SArray.jl]: Ah, righto.  Yes, I find the constructor/type dichotomy odd too.  I often wish the doc system would provide more help here, in terms of autodocumenting the type signature docs, in a way which is systematic but not intrusive.  Oh ho, look at this trick:

```julia
module A

"""
    $(B)
"""
type B{N,T,X,Y,Z}
end

end

help?> A.B
  A.B{N,T,X,Y,Z}

  B is a thing with some type parameters
```

Kinda close?

Comment (c42f) [inline src/SArray.jl]: Still, even if this was neater, it relies on package authors being systematic and good at writing docs...

File: src/SMatrix.jl (hunk 3 above)
Comment (c42f) [inline src/SMatrix.jl]: Sure.  Chuck in a TODO?

File: src/SizedArray.jl (hunk 5 above)
Comment (c42f) [inline src/SizedArray.jl]: heh :)

File: src/abstractarray.jl (hunk 6 above)
Comment (c42f) [inline src/abstractarray.jl]: Oh.  Duh!

File: src/SMatrix.jl (hunk 3 above)
Comment (andyferris) [inline src/SMatrix.jl]: issue?

**Pull request merged**